### ISSUE TITLE
feature request: stochastic computation graphs

### ISSUE THREAD

**Issue by predict-r** [apache/incubator-mxnet#2947]
It doesn't seem like custom gradients like this can be built automatically and is a nice idea. Could the log likelihood trick be implemented?

http://arxiv.org/pdf/1506.05254v3.pdf
http://blog.shakirm.com/2015/11/machine-learning-trick-of-the-day-5-log-derivative-trick/
https://www.youtube.com/watch?v=aUrX-rP_ss4
https://www.youtube.com/watch?v=oPGVsoBonLM
https://www.youtube.com/watch?v=rO7Dx8pSJQw
https://www.youtube.com/watch?v=gb5Q2XL5c8A
http://karpathy.github.io/2016/05/31/rl/

**Comment by piiswrong** [comment]
You can actually do it with mxnet. It's just not that automated.
You can chain up two mx.mod.module with the first ending in a softmax and the second's input being sampled according to the first ones output distribution

**Comment by predict-r** [comment]
When the gradient is calculated, it is different than the normal softmax gradient, so I don't see how. Please help.

On page 19 of the reinforcement learning course, Instead of multiplying the difference/delta/"score function" by results earlier in the chain, it is multiplied by the eventual rewards that might be sampled. The cost function is also based on these rewards.

While I hoped to use standard loss and gradients, this uses a log-likelihood trick to optimize over two things at once.

PAGE 19:
http://www0.cs.ucl.ac.uk/staff/D.Silver/web/Teaching_files/pg.pdf

**Comment by piiswrong** [comment]
MXNet support mixing graph and direct tensor computation. You can use symbol to do the standard part. Then you can do whatever you want as you would do in matlab.

**Comment by predict-r** [comment]
Thank you for the explanation. As you explained, I only need is to customize the final layer's gradient. However, I am doing it in R and unfortunately, mx.mod.module is for python.

**Comment by rudaoshi** [comment]
hi, @predict-r ，do you have implemented the stochastic computation graphs algorithm with mxnet or any thing else?

**Comment by predict-r** [comment]
not the computation graph. I was working on reinforcement learning. A computation graph is implemented efficiently in mxnet already, which is why I want to master it. The other techniques such as using twitter autograd are very good too, but it seems that at least one twitter autograd developer complements the mxnet team for efficiency, especially memory efficiency.

**Comment by sxjscience** [comment]
@predict-r I've done this before via custom operators. You can sample a value in the forward pass and calculate the backward gradient as reward (or reward minus baseline) times the gradient over the loglikelihood. @peterzcc and I are trying to rewrite these stochastic operators using C++.

**Comment by predict-r** [comment]
As we say in California, "cool"! Will you share your custom operators code in R or anything else? I've implemented this for the game connect4 using no special library and it seemed to be working, but it was my first version (see code near the end for the special loss function) `# source('~/Dropbox/Hot folder/reinforcement/mxnet/diy_connect4.R')
# game settings
test = FALSE
#batch_schedule = c(1000,200) #initial rollout and then smaller batches

batch_schedule = c(2000,rep(1000,ifelse(test,0,300))) #initial rollout and then smaller batches

# function names inspired by openai gym
players = c(1,2)

environment.reset = function(){
  matrix(0,6,7)
}
environment.action_space <- function(env){
  return(which(env[1,]==0))
}

action_space.sample <- function(ac){
  as.integer(sample(as.character(ac),1))
}

environment.timestep_limit = function(env){
  return(length(which(env==0)))
}

environment.render <- function(env){
  print(env)
  return(env)
}

environment.step = function(env_t0,player,action_t1){
  #just take an action and return new environment, nothing faccy
  open= which(env_t0[,action_t1]==0) #vector of open slots in column
  i = open[length(open)]
  env_t0[i,action_t1] = player
  return(env_t0)
}




done <- function(reward){
  ifelse(is.na(reward),FALSE,TRUE)
}

reward <- function(env,player){
  if(environment.timestep_limit(env)==0){
    #if no more spaces left, return 0
    return(-1)
  }
  lines = c( 
    apply(env,1,paste, collapse=''),
    apply(env,2,paste, collapse=''),
    paste(diag(env[3:6,]), collapse=''),
    paste(diag(env[2:6,]), collapse=''),
    paste(diag(env), collapse=''),
    paste(diag(env[,2:7]), collapse=''),
    paste(diag(env[,3:7]), collapse=''),
    paste(diag(env[,4:7]), collapse='')
  )
  if(any(grepl("1111",lines))){
    return(1)
  } else if(any(grepl("2222",lines))){
    return(-1)
  } else {
    return(NA)
  }
  
}

environment.size <- function(env){
  #number of numeric vectors to minimally describe environment state
  # include env at t0 and action at t1
  # 3 values for each space (empty, player1, player2)
  # 6*7 board 
  # 7 moves
  # 3*6*7 + 7 = 133
  # 3*6*7 = 126
  return(126 )
}


environment.unravel <- function(env_t0){
  unravel = integer()
  board=  c(
    as.integer(as.vector(env_t0)==0),
    as.integer(as.vector(env_t0)==1),
    as.integer(as.vector(env_t0)==2)
  )
 
  return( matrix(board,1,length(board)) )
}

environment.size <- function(env){
  #number of numeric vectors to minimally describe environment state
  # include env at t0 and action at t1
  # 3 values for each space (empty, player1, player2)
  # 6*7 board 
  # 7 moves
  return(3*6*7)
}





win_history = 0



# neural network
##################  
hidden = h = 200
D = environment.size()
K = length(environment.action_space(environment.reset()))


W = list()
b = list()
W[[1]]= matrix(rnorm(D*h, sd=0.01),D,h)
b[[1]] = rep(0,h)
W[[2]]= matrix(rnorm(h*K, sd=0.01),h ,K  )
b[[2]] = rep(0,K)


predict.diy <- function(W,b,X){
  hidden_layer = X %*% W[[1]] + matrix(b[[1]],nrow(X),ncol(W[[1]]),byrow=T) 
  hidden_layer[hidden_layer<0] <-0
  scores = hidden_layer %*% W[[2]] + matrix(b[[2]],nrow(hidden_layer),ncol(W[[2]]),byrow=T) 
  #predicted_class = apply(scores,1, which.max)
  exp_scores = exp(scores)
  probs = exp_scores/rowSums(exp_scores)
  return(probs)
}

step_size  = 0.6
reg = 1e-3



### batches ###

for(batch in 1:length(batch_schedule)){
  cat("\n[BATCH",batch, "]" )
  #get data ready for next minibatch
  X = NULL
  y = NULL
  REWARD = NULL
  # game
  
#  rewards = 0 # rewards across batches
  for(g in 1:batch_schedule[batch]){
    
    
    env = environment.reset()
    #since this game is always winnable if going first, learn going first
    players_turn = 1
    # players_turn = sample(players,1)
    player_1_actions = 0
    X_tmp = NULL
    y_tmp = NULL
    
    for(t in 1:environment.timestep_limit(env)){
      
      
      
      
      if(players_turn==1){
        
        #greedy epsilon for player 1
        #  sample(1:10,1)<=1 | 
        if((batch==1|sample(1:10,1)<=1) & !test ){
         
          action = action_space.sample(environment.action_space(env))
        } else {
         
          
            #act greedy and take best hypothesized action
            possible_plays = environment.action_space(env)
            newdata = environment.unravel(env)
            hypothesis = predict.diy(W,b,newdata)
            hypothesis[!1:7 %in% possible_plays] <- 0
             best = apply(hypothesis,1, which.max)
            action = action_space.sample(best)
          
         
          
        }
        #remember X's actions in a given env
        player_1_actions = player_1_actions +1
        X_tmp = rbind(X_tmp,environment.unravel(env))
        y_tmp = c(y_tmp,action)
        
        
      }else if(players_turn==2){
        # player 2's turn is always random
        action = action_space.sample(environment.action_space(env))
        
      }
      
      # given env, action, and players turn : 
      env = environment.step(env, player=players_turn, action)
      #switch players
      players_turn = which(players!=players_turn)[1]
      
      
      this_reward = reward(env)
      #episodic reward
      if(done(this_reward)){
        X = rbind(X, X_tmp)
        y = c(y,y_tmp)
        REWARD = c(REWARD,rep(as.integer(this_reward),player_1_actions))
       # rewards = rewards + this_reward
       # cat(paste("\n[BATCH ",batch,"REWARD AVERAGE] ",round(mean(REWARD)), collapse=''))
        break;
      }
    }
    
    
    
    
    
  } # end of particular batch of games g
  
  
   # learn policy
  if(!test){
    cat("\nTRAINING: ")
  if(batch==1){ epochs = 20} else {epochs =1} 
  for(epoch in 1:epochs){
    
  
    num_examples = nrow(X)
    hidden_layer = X %*% W[[1]] + matrix(b[[1]],nrow(X),ncol(W[[1]]),byrow=T) 
    hidden_layer[hidden_layer<0] <-0
    
    scores = hidden_layer %*% W[[2]] + matrix(b[[2]],nrow(hidden_layer),ncol(W[[2]]),byrow=T) 
   # predicted_class = apply(scores,1, which.max)
    
    exp_scores = exp(scores)
    probs = exp_scores/rowSums(exp_scores)
    
    # SPECIAL UPDATES
     baseline = mean(win_history[max(1,batch-5):batch])
     baseline = 0
    dscores = probs
    for(row in 1:nrow(probs)){
      onesy = rep(0,K)
      onesy[y[row]]<-1
      dscores[row,] = (onesy - dscores[row,]) * (REWARD[row]-baseline)
    }
    dscores= dscores/num_examples
    
    dW2 = t(hidden_layer) %*% dscores
    db2 = colSums(dscores)
    dhidden = dscores %*% t(W[[2]])
    dhidden[hidden_layer<=0] <- 0
    dW = t(X) %*% dhidden
    db = colSums(dhidden)
    
    dW2 = dW2 + reg * W[[2]]
    dW = dW + reg * W[[1]]
    
    W[[1]] = W[[1]] - step_size * dW
    b[[1]] = b[[1]] - step_size *db
    W[[2]] = W[[2]] - step_size * dW2
    b[[2]] = b[[2]] - step_size * db2
   cat("!")
    
  }
  }
  win_history = c(win_history,mean(REWARD))
  cat("\n[BATCH",batch,"MOVING AVRAGE WINS]:",round((mean(win_history[max(1,batch-5):batch])+1)/2,2)*100,"%")
  
 # cat(paste("\n[BATCH",batch,"REWARD AVERAGE] ",round(mean(REWARD),6), collapse=''))
  plot(win_history, type="l", xlab="batch", ylab="E(REWARD)")
  
  
} # end of all batches


#cat("\nMoving average of win percentage:",round((mean(win_history[max(1,batch-5):batch])+1)/2,2)*100,"%")



`

**Comment by sxjscience** [comment]
@predict-r Here's our implemented LogSoftmaxPolicy operator, which is samples a number according to the multinomial distribution in the forward pass and pass back the gradient based on the loglikelihood loss.
```python
class LogSoftmaxPolicy(mx.operator.CustomOp):
    """Outputs actions generated by a multinomial distribution whose parameters are computed
    by applying softmax to the input.
    Loss function for backward operation is set to be
      \sum_i - log(multinomial(a_i| softmax(x_i)) * R_i - H(multinomial(a_i|softmax(x_i)))
    """
    def __init__(self, deterministic, implicit_backward, use_mask, entropy_regularization, grad_scale):
        super(LogSoftmaxPolicy, self).__init__()
        self.deterministic = deterministic
        self.implicit_backward = implicit_backward
        self.use_mask = use_mask
        self.entropy_regularization = entropy_regularization
        self.grad_scale = grad_scale

    def forward(self, is_train, req, in_data, out_data, aux):
        x = in_data[0].asnumpy()
        y_prob = out_data[1].asnumpy()
        if self.use_mask:
            mask = in_data[1].asnumpy()
            mask = (mask > 0).astype(numpy.float32)
            y_prob[:] = mask * numpy.exp(x - numpy.ma.array(x, mask=1 - mask).max(axis=1).data.reshape((x.shape[0], 1)))
        else:
            y_prob[:] = numpy.exp(x - x.max(axis=1).reshape((x.shape[0], 1)))
        y_prob /= y_prob.sum(axis=1).reshape((x.shape[0], 1))
        if self.deterministic:
            self.assign(out_data[0], req[0], nd.array(numpy.argmax(y_prob, axis=1)))
            self.assign(out_data[1], req[1], nd.array(y_prob))
        else:
            y_action = sample_categorical(prob=y_prob, rng=get_numpy_rng())
            self.assign(out_data[0], req[0], nd.array(y_action))
            self.assign(out_data[1], req[1], nd.array(y_prob))

    def backward(self, req, out_grad, in_data, out_data, in_grad, aux):
        if self.use_mask:
            score = in_data[2].asnumpy()
        else:
            score = in_data[1].asnumpy()
        if self.implicit_backward:
            action = out_data[0].asnumpy().astype(numpy.int32)
        else:
            action = in_data[2].asnumpy().astype(numpy.int32)
        prob = out_data[1].asnumpy()
        dx = in_grad[0].asnumpy()
        dx[:] = prob
        dx[numpy.arange(action.shape[0]), action] -= 1.0
        dx[:] *= score.reshape(score.shape[0], 1)
        dx[:] += self.entropy_regularization * (prob * (entropy(prob.T).reshape(prob.shape[0], 1)) +
                                                numpy.nan_to_num(prob * numpy.log(prob)))
        self.assign(in_grad[0], req[0], self.grad_scale * dx)


@mx.operator.register("LogSoftmaxPolicy")
class LogSoftmaxPolicyProp(mx.operator.CustomOpProp):
    def __init__(self, deterministic=False, implicit_backward=True, use_mask=False,
                 entropy_regularization=0.01, grad_scale=1.0):
        super(LogSoftmaxPolicyProp, self).__init__(need_top_grad=False)
        self.deterministic = safe_eval(deterministic)
        self.implicit_backward = safe_eval(implicit_backward)
        self.use_mask = safe_eval(use_mask)
        self.entropy_regularization = safe_eval(entropy_regularization)
        self.grad_scale = safe_eval(grad_scale)

    def list_arguments(self):
        if self.implicit_backward:
            if self.use_mask:
                return ['data', 'mask', 'score']
            else:
                return ['data', 'score']
        else:
            if self.use_mask:
                return ['data', 'mask', 'score', 'backward_action']
            else:
                return ['data', 'score', 'backward_action']

    def list_outputs(self):
        return ['action', 'prob']

    def infer_shape(self, in_shape):
        data_shape = in_shape[0]
        score_shape = (in_shape[0][0],)
        action_shape = (in_shape[0][0],)
        prob_shape = in_shape[0]
        if self.implicit_backward:
            if self.use_mask:
                mask_shape = in_shape[0]
                return [data_shape, mask_shape, score_shape], [action_shape, prob_shape], []
            else:
                return [data_shape, score_shape], [action_shape, prob_shape], []
        else:
            backward_action_shape = (in_shape[0][0],)
            if self.use_mask:
                mask_shape = in_shape[0]
                return [data_shape, mask_shape, score_shape, backward_action_shape], [action_shape, prob_shape], []
            else:
                return [data_shape, score_shape, backward_action_shape], [action_shape, prob_shape], []

    def create_operator(self, ctx, shapes, dtypes):
        return LogSoftmaxPolicy(deterministic=self.deterministic,
                                implicit_backward=self.implicit_backward,
                                use_mask=self.use_mask,
                                entropy_regularization=self.entropy_regularization,
                                grad_scale=self.grad_scale)
```

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
